sqlx-sqlite/src/index.rs

```diff
@@ -0,0 +1,599 @@
+/// | 1       | 2         | 100       | 0                  | 201         | 300         | 0                  | "a"         | "z"         |
+/// | 2       | 0         | 50        | 0                  | 1           | 100         | 0                  | "x"         | "x"         |
+/// | 2       | 1         | 100       | 0                  | 101         | 200         | 0                  | "y"         | "z"         |
+/// | 2       | 2         | 150       | 0                  | 201         | 300         | 0                  | "123"       | "456"       |
+///
+/// To do filtering we rewrite the filter expression to a set of SQL expressions that can be evaluated against the index.
+/// For example, if we have a filter `a = 5` we would rewrite that to `a_min <= 5 AND a_max >= 5`:
+/// 
+/// ```sql
+/// WITH row_groups AS (
+///   SELECT file_id, row_group
+///   FROM row_group_statistics
+///   WHERE a_min <= 5 AND a_max >= 5
+/// )
+/// SELECT file_name, file_size_bytes, row_group
+/// FROM file_statistics
+/// JOIN row_groups USING (file_id)
+/// ```
+///
+/// While we use SQLite in this example, the index could be implemented with other databases or system.
+/// SQLite is just a convenient example that is also very similar to other RDBMS systems that you might use.
+#[derive(Debug)]
+pub struct SQLiteIndex {
+    pool: SqlitePool,
+    /// The index for the schema. Not all columns in the table need to be indexed.
+    schema: SchemaRef,
+}
+
+impl Display for SQLiteIndex {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "SQLiteIndex()")?;
+        Ok(())
+    }
+}
+
+impl SQLiteIndex {
+    pub fn new(pool: SqlitePool, schema: SchemaRef) -> Self {
+        Self { pool, schema }
+    }
+
+    /// Return the filenames / row groups that match the filter
+    ///
+    /// This function pushes down the filter to the index to get a list of row groups that match
+    /// and hence the files that need to be read.
+    ///
+    /// The filter is pushed down to the index by converting it to a set of SQL expressions
+    /// that can be evaluated by the index.
+    ///
+    /// The return value is a list of `(file_name, FileScanPlan)` tuples where `FileScanPlan` contains
+    /// the file metadata and the row groups that need to be scanned.
+    pub async fn get_files(
+        &self,
+        filter: Arc<dyn PhysicalExpr>,
+        schema: SchemaRef,
+    ) -> Result<Vec<(String, FileScanPlan)>> {
+        // Convert the predicate to a pruning predicate
+        // This transforms e.g. `a = 5` to `a_min <= 5 AND a_max >= 5`
+        let pruning = PruningPredicate::try_new(filter, schema.clone())?;
+        let predicate = pruning.predicate_expr().clone();
+        // Replace any `{col}_row_count` with `row_count` as we don't store per-column row counts
+        // (they're the same for all columns in a row group)
```

Comment:
Where does `{col}_row_count` come from?

Reply:
PruningPredicate generates it. It checks for `{col}_row_count == `{col}_null_count`.

Reply:
1073e81